src/hotspot/cpu/riscv/assembler_riscv.cpp

```diff
@@ -352,20 +352,35 @@ void Assembler::li(Register d, void* addr) {
 void Assembler::li(Register d, long imm) { // TODO optimize
   unsigned long uimm = imm;
 
+  if (imm == 0) {
+      addi(d, R0_ZERO, 0);
+      return;
+  }
+
   // Accurate copying by 11 bits.
-  int remBit = ((uimm & 0xffffffff00000000) != 0) ? 53 : 21;
-  short part = (uimm >> remBit) & 0x7FF;
-  addi(d, R0_ZERO, part);
-  slli(d, d, 11);
-
-  for (remBit -= 11; remBit > 0; remBit -= 11) {
-    part = (uimm >> remBit) & 0x7FF;
-    if (part != 0) addi(d, d, part);
-    slli(d, d, remBit >= 11 ? 11 : remBit);
+  int remBit = ((uimm & 0xffffffff00000000) != 0) ? 55 : 22;
```

Comment:
Canonical way to construct 32-bit constant is `lui` + `addi`. Can we use it at least for such consts?
Please don't forget adjusting `lui` value if bit 12 is set. I think you saw this, but anyway http://kspt.icc.spbstu.ru/media/files/2018/lowlevelprog/riscv_isa_2.pdf
-- slide21: "LUI and AUIPC instructions **(3)**"

I cannot find an easy way to get 64-bit consts. For future investigation, probably, most of them are runtime addresses, than can be computed by pc-relative math. This may require reviewing `load_const` uses